In [1]:
# importng relevant library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Loading the data
df = pd.read_csv("Heart disease.csv")
# checking head
df.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   
5   57    1   0       140   192    0        1      148      0      0.4      1   
6   56    0   1       140   294    0        0      153      0      1.3      1   
7   44    1   1       120   263    0        1      173      0      0.0      2   
8   52    1   2       172   199    1        1      162      0      0.5      2   
9   57    1   2       150   168    0        1      174      0      1.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  
5   0     1       1  
6   0     2       1  
7   0     3       1  
8   0     3       1  
9   0     2       1

In [3]:
# Info of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [4]:
# checking missing value
df.isnull().sum() # no missing value so far

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [4]:
# checking unique value for all columns
df.nunique()

age          41
sex           2
cp            4
trestbps     49
chol        152
fbs           2
restecg       3
thalach      91
exang         2
oldpeak      40
slope         3
ca            5
thal          4
target        2
dtype: int64

In [5]:
# describing the data
df.describe(include="all")

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

### Writing Description about every columns

[Article Found on the heart dieases data](https://towardsdatascience.com/exploratory-data-analysis-on-heart-disease-uci-data-set-ae129e47b323)

[Data Description on Kaggle](https://www.kaggle.com/cherngs/heart-disease-cleveland-uci)

[Heart Disease Analysis System Using Data Mining Techniques](https://www.rroij.com/open-access/heart-disease-analysis-system-using-datamining-techniques.php?aid=50600)



`age` - person age in number (int).

`sex` - gender of individual || 1=male, 0 = female.

`cp` - Chest pain type, 0: asymptomatic || 1: atypical angina, 2: non-anginal pain, 3: typical angina

`trestbps`- blood preasure reading in mm/hg

`chol` - Serum Cholesterol in mg/dl

`fps` -Fasting Blood Sugar > 120 mg/dl: [0 = no, 1 = yes]
    A fasting blood sugar level below 100 milligrams per deciliter (mg/dL) — 5.6 millimoles per liter (mmol/L) — is considered normal. A fasting blood sugar level from 100 to 125 mg/dL (5.6 to 7.0 mmol/L ) is considered prediabetes. This result is sometimes called impaired fasting glucose
    
`restecg` (Resting ECG): [0: showing probable or definite left ventricular hypertrophy by Estes’ criteria, 1: normal, 2: having ST-T wave abnormality]


`thalach`- (maximum heart rate achieved)

`exang` - (Exercise Induced Angina): [1 = yes, 0 = no]

`oldpeak` -ST depression induced by exercise relative to rest

`slope` -the slope of the peak exercise ST segment: [0: downsloping; 1: flat; 2: upsloping]

`ca` -  [number of major vessels (0–3)

`thal` : [1 = normal, 2 = fixed defect, 3 = reversible defect]

`target`: [0 = disease, 1 = no disease


* Binary: sex, fbs, exang, target
* Categorical: cp, restecg, slope, ca, thal
* Continuous: age, trestbps, chol, thalac, oldpeak

In [6]:
# fixing `ca` column
df['ca'].value_counts()

0    175
1     65
2     38
3     20
4      5
Name: ca, dtype: int64

In [7]:
# there are 4 rows having 4 in the data, but document says it should be range of 0 to 3
# droping those rows having 4 as value
i_4 = df[df['ca']==4].index
print(i_4)
df.drop(i_4, inplace=True) # droping all the columns having 4 values

Int64Index([92, 158, 163, 164, 251], dtype='int64')


In [8]:
# chek it worekd or not
df[df['ca']==4] # expected nothing to return.

Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, target]
Index: []

In [9]:
df['thal'].value_counts()

2    163
3    115
1     18
0      2
Name: thal, dtype: int64

In [10]:
# Documnet says `thal` range is 1 to 3
# remving ectra 2 rows where thal calues is 0
i_0 = df[df['thal']==0].index
i_0

Int64Index([48, 281], dtype='int64')

In [11]:
# removing those 2 columns
df.drop(i_0, inplace=True)
df[df['thal']==0] # except nothing to return

Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, target]
Index: []

In [12]:
# checking for duplicate rows
print(df.shape)
df[df.duplicated()]

(296, 14)


Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, target]
Index: []

In [ ]:
# df.drop(164, inplace=True) # removing duplicate row from the table
# print(df.shape) # Checking the shape

#### As not every columns has categorical label, we have to encoding all numerical values to categorical value


In [13]:
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [15]:
df['thalach'].max()

202

In [17]:
# encoding target column
df['target'] = df.target.replace({1: "Disease", 0: "No_disease"})

# encoding gender column
df['sex'] = df.sex.replace({1: "Male", 0: "Female"})

# encoding different type of chest pain
df['cp'] = df.cp.replace({0: "typical_angina", 
                          1: "atypical_angina", 
                          2:"non-anginal pain",
                          3: "asymtomatic"})

# encoding exang column
df['exang'] = df.exang.replace({1: "Yes", 0: "No"})

# encoding fasting blood sugar
df['fbs'] = df.fbs.replace({1: "True", 0: "False"})

# encoding slope column
df['slope'] = df.slope.replace({0: "upsloping", 1: "flat", 2: "downsloping"})

# encoding thal column
df['thal'] = df.thal.replace({1: "fixed_defect", 2: "reversable_defect", 3:"normal"})

# checking the data
df.head(10)

age     sex                cp  trestbps  chol    fbs  restecg  thalach  \
0   63    Male       asymtomatic       145   233   True        0      150   
1   37    Male  non-anginal pain       130   250  False        1      187   
2   41  Female   atypical_angina       130   204  False        0      172   
3   56    Male   atypical_angina       120   236  False        1      178   
4   57  Female    typical_angina       120   354  False        1      163   
5   57    Male    typical_angina       140   192  False        1      148   
6   56  Female   atypical_angina       140   294  False        0      153   
7   44    Male   atypical_angina       120   263  False        1      173   
8   52    Male  non-anginal pain       172   199   True        1      162   
9   57    Male  non-anginal pain       150   168  False        1      174   

  exang  oldpeak        slope  ca               thal   target  
0    No      2.3    upsloping   0       fixed_defect  Disease  
1    No      3.5    upsloping   0  reversable_defect  Disease  
2    No      1.4  downsloping   0  reversable_defect  Disease  
3    No      0.8  downsloping   0  reversable_defect  Disease  
4   Yes      0.6  downsloping   0  reversable_defect  Disease  
5    No      0.4         flat   0       fixed_defect  Disease  
6    No      1.3         flat   0  reversable_defect  Disease  
7    No      0.0  downsloping   0             normal  Disease  
8    No      0.5  downsloping   0             normal  Disease  
9    No      1.6  downsloping   0  reversable_defect  Disease

In [19]:
# saving the data to another file
df.to_csv("Heart_Disease_cat_encode.csv", index=False)

In [21]:
# chcecking the saved file
df2 = pd.read_csv("Heart_Disease_cat_encode.csv")
df2.head(10)

age     sex                cp  trestbps  chol    fbs  restecg  thalach  \
0   63    Male       asymtomatic       145   233   True        0      150   
1   37    Male  non-anginal pain       130   250  False        1      187   
2   41  Female   atypical_angina       130   204  False        0      172   
3   56    Male   atypical_angina       120   236  False        1      178   
4   57  Female    typical_angina       120   354  False        1      163   
5   57    Male    typical_angina       140   192  False        1      148   
6   56  Female   atypical_angina       140   294  False        0      153   
7   44    Male   atypical_angina       120   263  False        1      173   
8   52    Male  non-anginal pain       172   199   True        1      162   
9   57    Male  non-anginal pain       150   168  False        1      174   

  exang  oldpeak        slope  ca               thal   target  
0    No      2.3    upsloping   0       fixed_defect  Disease  
1    No      3.5    upsloping   0  reversable_defect  Disease  
2    No      1.4  downsloping   0  reversable_defect  Disease  
3    No      0.8  downsloping   0  reversable_defect  Disease  
4   Yes      0.6  downsloping   0  reversable_defect  Disease  
5    No      0.4         flat   0       fixed_defect  Disease  
6    No      1.3         flat   0  reversable_defect  Disease  
7    No      0.0  downsloping   0             normal  Disease  
8    No      0.5  downsloping   0             normal  Disease  
9    No      1.6  downsloping   0  reversable_defect  Disease

In [22]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296 entries, 0 to 295
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       296 non-null    int64  
 1   sex       296 non-null    object 
 2   cp        296 non-null    object 
 3   trestbps  296 non-null    int64  
 4   chol      296 non-null    int64  
 5   fbs       296 non-null    bool   
 6   restecg   296 non-null    int64  
 7   thalach   296 non-null    int64  
 8   exang     296 non-null    object 
 9   oldpeak   296 non-null    float64
 10  slope     296 non-null    object 
 11  ca        296 non-null    int64  
 12  thal      296 non-null    object 
 13  target    296 non-null    object 
dtypes: bool(1), float64(1), int64(6), object(6)
memory usage: 30.5+ KB
